core/src/main/java/org/apache/druid/math/expr/ApplyFunction.java

```diff
@@ -105,6 +103,44 @@ default boolean hasArrayOutput(LambdaExpr lambdaExpr)
    */
   void validateArguments(LambdaExpr lambdaExpr, List<Expr> args);
 
+  /**
+   * Check that the argument list is the expected size. The list of arguments should not include the
+   * {@link LambdaExpr} and so the error message will indicate that count + 1 arguments are required.
+   */
+  default void validateArgumentCount(LambdaExpr lambdaExpr, List<Expr> args, int count)
```

Comment:
I am not sure if `validateArgumentCount` and `validateMinArgCount` should be a part of the interface `ApplyFunction`. They should probably just be static utility methods.

It makes sense to have `validateArguments` in the interface because it is a part of the general behaviour of a function to check if a given set of args is valid for that function. But validating argument count is probably too specific. For example, if in the future, we decide to validate other things, say the type of each of the args, etc, we would have to add another `validateArgumentType` method.

Reply:
These are helper methods to help things implement `validateArguments`. I went with this to cut down on the number of arguments that callers need to pass to them. To try to clarify the role of these methods, I have renamed them to things like `validationHelperCheckArgumentCount` and added additional javadocs

Reply:
actually, thinking a bit more on this I'm going to look into making a common interface for `Function`, `ApplyFunction`, and `ExprMacro`, since they are all doing the same sort of role

Reply:
reworked things a bit, added a new interface `NamedFunction`, which `Function`, `ApplyFunction` and `ExprMacro` all extend and which defines all of these validation helper methods so they can be shared